### DISCUSSION TITLE
Reusing structs in models

### DISCUSSION THREAD

**Discussion by ohuu** [diesel-rs/diesel#3874] [Q&A] [answered]
Not sure if this has already been asked. I've had a look around and can't find an answer to it, mostly because I'm not sure what to search for. Apologies if this i an old question!

It seems there's a lot of boilerplate when defining models in diesel. For example if I have a type with an `id` which is auto-incremented then you end up with something like this:

```rust
#[derive(Queryable, Selectable)]
#[diesel(table_name = crate::schema::post)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct Post {
    pub id: i32,
    pub post_title: String,
    pub post_body: String,
    ...
}

#[derive(Insertable)]
#[diesel(table_name = crate::schema::post)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct NewPost {
    pub post_title: String,
    pub post_body: String,
    ...
}
```

in the spirit of DRYness I'd like to be able to factor out `post_title` and `post_body` into another struct and use it in both `Post` and `NewPost`. Is this possible?

**Comment by weiznich** [comment, marked as answer]
You can just put these fields in a seperate struct and use that struct as field in your `Post` and `NewPost` types via the `#[diesel(embed)]` attribute offered by [`#[derive(Insertable)]`](https://docs.diesel.rs/2.1.x/diesel/prelude/derive.Insertable.html#optional-field-attributes) and [`#[derive(Selectable)]`](https://docs.diesel.rs/2.1.x/diesel/expression/derive.Selectable.html#field-attributes)

```rust
#[derive(Queryable, Selectable)]
#[diesel(table_name = crate::schema::post)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct Post {
    pub id: i32,
    #[diesel(embed)]
    fields: SharedPost
}

#[derive(Insertable)]
#[diesel(table_name = crate::schema::post)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct NewPost {
    #[diesel(embed)]
     fields: SharedPost
}

#[derive(Insertable, Selectable, Queryable)]
#[diesel(table_name = crate::schema::post)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct SharedPost {
    pub post_title: String,
    pub post_body: String,
}
```

If you are using `serde` for (de)serializing the result data the `#[serde(flatten)]` attribute is really helpful here to keep the flat json structure to the outside.
